### PR TITLE
feat(recipes): add Qwen3.6 27B DPO and GRPO

### LINKED ISSUE #845: Recipe: qwen3.6-27b-dpo and qwen3.6-27b-grpo (part of #275)
## What

Add two recipes to `src/soup_cli/recipes/catalog.py`:

- `qwen3.6-27b-dpo`
- `qwen3.6-27b-grpo`

`Qwen/Qwen3.6-27B` currently ships SFT only (`qwen3.6-27b-sft`). This is one row of the #275 tracker, split out so it can be picked up on its own. **Comment on this issue before you start** so two people do not build it.

## Base repo

`Qwen/Qwen3.6-27B` — checked on 2026-09-11 with `huggingface_hub.model_info(..., token=False)`: it resolves, it is not gated, and its config reports `model_type: qwen3_5`. Please re-check it yourself the same way. Do **not** use an HTTP status as evidence: an invented repo id returns the same 401 as a private one, which is how #661 happened. `scripts/check_recipe_repo_ids.py` does this correctly.

## What to copy, and from where

Start from the sibling `qwen3.6-27b-sft` (dense, `modality: text`, `max_length: 4096`, LoRA r=16 / alpha=32, `quantization: 4bit`, `gradient_accumulation_steps: 8`) and swap in the task-specific block from the closest template:

| recipe | task block from | fields that change |
|---|---|---|
| `qwen3.6-27b-dpo` | `qwen3.5-9b-dpo` | `task: dpo`, `data.train: ./data/preference_train.jsonl`, `data.format: dpo`, `lr: 5e-6`, `dpo_beta: 0.1` |
| `qwen3.6-27b-grpo` | `qwen3.5-9b-grpo` | `task: grpo`, `data.train: ./data/reasoning_train.jsonl`, `grpo_beta: 0.1`, `num_generations: 4`, `reward_fn: accuracy` |

The catalog's own numbers, so you can justify values with counts rather than "matches the others": across all 14 DPO recipes, `lr: 5e-6` holds 14/14 and `dpo_beta: 0.1` holds 14/14; across all 23 GRPO recipes, `lr: 1e-5` holds 16/23 and `num_generations: 4` holds 23/23. If you deviate from the sibling (for example GRPO's batch shape), say so in the PR body before review asks.

A quick sanity check: YAML assembled exactly as in the table parses through `load_config_from_string` on current `main` for both tasks. That proves the schema accepts it, nothing more.

## The ratchets that must pass

- [ ] `RecipeMeta(...)` entries in `src/soup_cli/recipes/catalog.py`, tags/description following the sibling.
- [ ] `modality: text` in both YAMLs, and both names added to `EXPECTED_QWEN35_TEXT_RECIPES` in `tests/test_issue427_qwen35_text_modality.py:11`. The family audit at `:43` fails on any Qwen3.5/3.6/3.8 recipe that is not registered, and the checkpoint's architecture is multimodal (#427), so the text-only intent has to be explicit.
- [ ] Model id pinned on **two independent surfaces**, each by its own test: `RecipeMeta.model` and the YAML `base:`. Mirror `TestDeepSeekV4FlashDpoRecipe` in `tests/test_recipes.py:503`. The snapshot guard only sees the YAML-derived config, so a wrong `RecipeMeta.model` passes everything else.
- [ ] Catalog count bumped everywhere (165 today, plus one per recipe; use whatever the number is when you rebase):
  - `src/soup_cli/recipes/catalog.py:51` comment;
  - `test_catalog_size_is_N` renamed and re-asserted in `tests/test_recipes.py:936`, `tests/test_v07130.py:740`, `tests/test_v07132.py:953`, plus the history line near `tests/test_recipes.py:964`;
  - `tests/test_v07124.py:281` assert, and the docstring at `:277` that names the test **as a string** — nothing catches that one if you miss it;
  - the four doc sites checked by `tests/test_recipe_count_is_synced.py:83-96` (`CONTRIBUTING.md`, `docs/commands.md`, `docs/serving-and-export.md` twice).
- [ ] `python scripts/generate_recipe_snapshot.py` run **last**, after every count edit, and the diff to `tests/fixtures/recipe_config_snapshots.json` is purely additive (your recipe blocks, zero deletions).
- [ ] A changelog fragment `changelog.d/0.74.0/<PR>.added.md` (see `changelog.d/README.md`; use the PR number, not this issue's).
- [ ] `soup recipes show <name>` and `soup recipes use <name>` run live; paste the output in the PR.

## A note on mutations

Some values you write equal the schema default (`dpo_beta: 0.1`, `grpo_beta: 0.1`, `num_generations: 4`, `reward_fn: accuracy`, `epochs: 3`, `batch_size: auto`). Deleting such a line yields a byte-identical resolved config, so no test can kill it. Report those as survivors in the PR body rather than inventing a test to kill them. The distinctive lines — model id, task, data format, `lr`, `modality` — are the ones worth pinning.

## Honest limits

A 27B model will not train on a consumer card, and nobody reviewing this expects you to. What you can and should verify: the entry loads, both id surfaces are pinned, show/use work, the count sweep is complete, the snapshot diff is additive. The hyperparameters are inherited defaults, not a tuned recommendation — the PR description should say which line came from which template.

### PR DESCRIPTION
**Pull request by akkupratap323**
Qwen3.6-27B shipped with only an SFT recipe, leaving the DPO and GRPO task variants from #275 unavailable. This adds both dense 27B variants, preserving the sibling's explicit text modality, 4-bit LoRA shape, and `auto` / 8 batch settings while applying the established preference and reasoning task blocks.

Both `RecipeMeta.model` and YAML `base:` are pinned independently. `Qwen/Qwen3.6-27B` was re-checked with `huggingface_hub.model_info(..., token=False)` and resolves public, ungated, with `model_type=qwen3_5` and `pipeline_tag=image-text-to-text`; both recipes therefore join the explicit Qwen text-tower guard.

Catalog count moves 174 -> 176 across every live count site. Historical count records remain unchanged, and the regenerated snapshot adds only the two new keys.

Validation:

- 347 passed across recipe, modality, count-sync, and snapshot suites
- 2 passed for the v0.71.x catalog-size assertions
- `ruff check src/soup_cli tests`
- `soup recipes show` and `soup recipes use` passed for both recipes; both generated files reload through `load_config_from_string`

The broader v0.71.30 file has 18 unrelated local failures because optional `torch` and `safetensors` are absent; its catalog-size test passes. The 27B configurations inherit catalog defaults and were not trained locally.

Fixes #845

### CODE REVIEW

**Comment by MakazhanAlpamys** [comment]
**Approved on substance.** Two cosmetic fixes before I merge, and the first one is the repo's own documented mistake repeating.

**Your nine red cells are #1005, not you — I read the logs rather than assuming.** All nine cells, from the job endpoint: `test_v07201.py` (5), `test_v07202.py` (2-3), `test_v07203.py` (1). The ubuntu-vs-windows/macos 9-vs-8 delta is the `#382` `_windows_ci` skip and its macOS equivalent, which is exactly right. **No tenth failure in any cell.** `lint`, `type-check`, `pytorch-smoke`, `transformers-floor`, `mlx-smoke`, `validate-recipes` all pass. #1010 has since merged as `f8226214`, so a re-run will be clean.

## Verified by derivation, not from your description

`RecipeMeta(` entries: **174** on `origin/main`, **176** on your head. `len(RECIPES)` imported from a worktree: **176**. Matches your 174 -> 176.

The snapshot is a **clean regeneration** — re-running `scripts/generate_recipe_snapshot.py` wrote 299 baseline fields and 176 deltas, after which `git status --porcelain` was empty. Not a hand-edit.

**All 176 recipes still load** through the real `load_config_from_string`, zero failures. That check matters more than usual here: #983 recently added a config-load refusal for `use_flash_attn` / `use_liger` outside `{sft, tts}`, and you are adding a **DPO** and a **GRPO** recipe — the exact collision shape, which git reports no conflict for. Neither of yours sets either flag.

`model_info("Qwen/Qwen3.6-27B", token=False)` resolves, public and ungated, `pipeline_tag=image-text-to-text`. That tag is precisely why both belong in `EXPECTED_QWEN35_TEXT_RECIPES` and why both YAMLs carry an explicit `modality: text` — consistent with the `qwen3.6-27b-sft` sibling. You got that right.

**The two-surface pin holds, and it reproduces #978's asymmetry exactly.** Anchor `base: Qwen/Qwen3.6-27B` + `task: dpo` — matched **once** (the bare `base:` line appears three times, so the task line is what makes it unique). Mutating the YAML's `base:` while leaving `RecipeMeta(model=...)` intact, module import confirmed (176 recipes still load):

```
FAILED test_recipes.py::TestQwen36DenseTaskVariantRecipes::test_yaml_base_pins_the_model_id
FAILED test_recipes.py::TestQwen36DenseTaskVariantRecipes::test_recipe_loads_with_expected_dense_shape
FAILED test_issue621_recipe_snapshot.py::...::test_recipe_delta_matches_its_snapshot[qwen3.6-27b-dpo]
3 failed, 344 passed
```

and `test_recipe_meta_pins_the_model_id` **passed** under that same mutation. That asymmetry is the point rather than a gap: the `RecipeMeta` surface is blind to the YAML, so only a test that parses the YAML can catch the two drifting apart.

Deleted lines: five, every one the ratchet moving 174 -> 176. `test_issue427_qwen35_text_modality.py` gains two lines and loses none. 377 + 366 + 21 passed across the affected files; ruff clean. Fragment is in the right directory, named after the **PR**, and credits you.

---

## Finding 1 — two tests are now named `..._is_174` while asserting `176`, and this repo has already written the post-mortem on that

You renamed `test_recipes.py`'s copy correctly and missed two:

- `tests/test_v07130.py:740` — `test_catalog_size_is_174`, asserts 176
- `tests/test_v07132.py:953` — `test_catalog_size_is_174`, asserts 176

`tests/test_v07124.py:270-278` carries the previous occurrence verbatim:

> *"Renamed 2026-09-02: the old name said 158 while asserting 162. It had drifted through 116 -> 134 -> 137 -> 158 -> 162 without the name following, so the one thing a reader saw first was the one thing that was false."*

Nothing fails, which is exactly why it drifts. Rename both to `_is_176`.

## Finding 2 — a cross-reference you broke by fixing the thing it points at

`tests/test_v07124.py:277` says *"The count itself is pinned by `test_catalog_size_is_174` in `test_recipes.py`"* — and you renamed that test to `_is_176`. The name it cites no longer exists.

Better than updating the number: **drop the number from the reference** entirely, since it will rot again next release.

## Not your problem, recorded so it is not mistaken for one

Thirteen sites for one integer is why three drift on every recipe PR — `docs/serving-and-export.md` alone holds it twice, and `test_v07130`/`test_v07132` each carry a duplicate of an assertion `test_recipe_count_is_synced.py` already owns. I will file that as a maintainer issue; **do not fix it here.**

Push the two renames and the cross-reference and I will merge.

**Comment by MakazhanAlpamys** [comment]
**The issue I promised you is filed: #1016.** Paying that before nudging you again, since it was my debt and it came out of your review.

It credits what you actually did rather than describing the problem abstractly: you rebased #978 after #956 landed `171 -> 172` underneath you and re-edited every site by hand without complaint, and that is the concrete cost the issue argues from. I also wrote down the two things the fix must **not** do — delete the milestone assertions (the `RecipeMeta` surface is blind to the YAML, which your own two-surface pin demonstrated), and touch `README.md`'s deliberate `100+`.

For the record, the count is nine sites on `main` at `f8226214`, not the thirteen I said to you earlier. I had been counting the internal `.claude/history/*` files and the narrating comment in `test_recipes.py:1615-1617` alongside the real ones. Nine is the number a contributor actually has to edit; the `.claude/` ones are gitignored and are my problem, not yours.

Your PR is still waiting on the two renames and the cross-reference. No new findings from me — head unchanged, so this is not a re-review.

**Comment by MakazhanAlpamys** [comment]
**Approved. I merge the moment the required checks are green on `3a1399b3`** — all three asks are satisfied, and one of them in the stronger form rather than the one I literally asked for.

Turned around in seven minutes, which I noticed.

## The three, verbatim

| ask | status |
|---|---|
| `test_v07130.py:740` — rename `test_catalog_size_is_174` to `_is_176` | done, asserts 176 |
| `test_v07132.py:953` — same | done, asserts 176 |
| `test_v07124.py:277` — **drop the number** from the cross-reference rather than update it | done, **dropped** |

The third is the one worth calling out. I said the number would rot again next release and asked you to remove it rather than bump it, and you removed it — the docstring now reads "pinned in `test_recipes.py` and by `test_recipe_count_is_synced.py`" with no digit at all. That is the version that stops costing someone an edit every release.

Repo-wide grep over tracked files for `_is_174`, `== 174` and `174 recipes`: **zero hits.** The only `_is_N` names left are the three legitimately renamed to `_is_176`.

## What I verified rather than took on trust

`len(RECIPES)` derived on the merged tree, with `PYTHONPATH` pinned into the checkout and the pin proved before every run — this matters acutely here, because an unpinned import would have read the catalog out of the stale v0.75.0 copy in `site-packages` and reported a confidently wrong count.

**176, and all nine sites agree** — `catalog.py:51`, `CONTRIBUTING.md:126`, `docs/commands.md:155`, `docs/serving-and-export.md:551` and `:560`, and the four test assertions. Nothing left behind.

All **176** recipes load through the real `load_config_from_string`, zero failures, and for every one `cfg.base == meta.model` and `cfg.task == meta.task`. Your two:

```
qwen3.6-27b-dpo   base=Qwen/Qwen3.6-27B  task=dpo   quant=4bit  lora.r=16
qwen3.6-27b-grpo  base=Qwen/Qwen3.6-27B  task=grpo  quant=4bit  lora.r=16
```

The repo ID is **character-identical** to the `qwen3.6-27b-sft` sibling already in the catalog, so no #661-shaped typo. Neither sets `use_flash_attn` or `use_liger`, so the #983 collision I was worried about is confirmed absent by execution rather than by git staying quiet.

**Three mutations, all killed, each anchor matched exactly once with module import confirmed:**

- YAML `base:` mutated with `RecipeMeta model=` left intact → **3 failed**, including the snapshot guard naming the field pair: `{'base': ('Qwen/Qwen3.6-27B', 'Qwen/Qwen3.6-27B-MUTANT')}`. And the asymmetry holds — `test_recipe_meta_pins_the_model_id` **passes** under that same mutation, because the `RecipeMeta` surface is blind to the YAML. That is the design, not a gap, and it is why the YAML-parsing test has to exist separately.
- `RecipeMeta model=` mutated with the YAML intact → **3 failed**, including `assert {"sft","dpo","grpo"} <= tasks` reporting `Extra items in the left set: 'dpo'`.
- one recipe removed from `RECIPES` entirely → **13 failed**, and the ratchet message names every out-of-sync site with its file, line and regex. That is the message a first-time contributor actually needs.

Nine removed test lines, every one accounted for: three renames, four ratchet bumps `174 -> 176`, two lines of the docstring reflow I asked for. No test function removed, nothing weakened, no skip added. The snapshot fixture is a **clean regeneration** — I re-ran `scripts/generate_recipe_snapshot.py` and `git status` came back empty, so it is not a hand-edit.

`ruff` clean. `test_recipes.py` 134 passed, `test_recipe_count_is_synced.py` 10 passed, `test_issue621_recipe_snapshot.py` 180 passed, plus 419 more across the touched milestone files.

## Two notes, neither needing a push

1. The `test_v07124.py` docstring reflow leaves one short ragged line. ruff-format does not reflow docstrings so it is clean; it is just not re-wrapped. Not worth a round trip.
2. Your force-push rebased onto `f8226214`, which is why your CI is not showing the nine `test_v07201/2/3` failures that several other open PRs are currently showing. Those are #1005, fixed by #1010 this morning; PRs whose runs predate it are stuck with a red X they cannot clear without pushing. Your rebase sidestepped it by accident of timing.

## And the issue I owed you

**#1016** is filed — the nine-sites problem, argued from the concrete cost rather than in the abstract, with your #978 rebase after #956 landed `171 -> 172` underneath you as the lead example. It also records the two things a fix must not do: delete the milestone assertions (your two-surface pin above is exactly why that surface still needs pinning), and touch `README.md`'s deliberate `100+`.

Thank you for this one. The count discipline across nine sites is tedious and invisible when it goes right, and you have now done it twice without being chased.

**Comment by MakazhanAlpamys** [comment]
Merged as 31b7d04e. Thank you, @akkupratap323.

All three review asks landed, and you did the cross-reference the stronger way: you dropped the number instead of bumping it, so it cannot go stale again. On the tree merged with current `main`, all 176 recipes load, the count agrees at all nine sites, and the config snapshot is a clean regeneration. That run was F821 clean, with 2393 passed across the recipe, snapshot, unknown-key (#627) and consumer-guard (#748) tests. You also wrote down in the fragment that the 27B configs were validated but not trained, which is exactly what a reader needs to know.
